hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/AuthAwareTokenConverter.kt

```diff
@@ -0,0 +1,61 @@
+package uk.gov.justice.hmpps.kotlin.auth
+
+import org.springframework.core.convert.converter.Converter
+import org.springframework.security.authentication.AbstractAuthenticationToken
+import org.springframework.security.core.GrantedAuthority
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.springframework.security.oauth2.jwt.Jwt
+import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
+import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter
+import reactor.core.publisher.Mono
+
+class AuthAwareTokenConverter : Converter<Jwt, AbstractAuthenticationToken> {
+  private val jwtGrantedAuthoritiesConverter: Converter<Jwt, Collection<GrantedAuthority>> =
+    JwtGrantedAuthoritiesConverter()
+
+  override fun convert(jwt: Jwt) = convert(jwt, jwtGrantedAuthoritiesConverter)
+}
+
+class AuthAwareReactiveTokenConverter : Converter<Jwt, Mono<AuthAwareAuthenticationToken>> {
+  private val jwtGrantedAuthoritiesConverter: Converter<Jwt, Collection<GrantedAuthority>> =
+    JwtGrantedAuthoritiesConverter()
+
+  override fun convert(jwt: Jwt): Mono<AuthAwareAuthenticationToken> = Mono.just(convert(jwt, jwtGrantedAuthoritiesConverter))
+}
+
+private fun convert(jwt: Jwt, converter: Converter<Jwt, Collection<GrantedAuthority>>): AuthAwareAuthenticationToken {
+  val claims = jwt.claims
+  val principal = findPrincipal(claims)
+  val authorities = extractAuthorities(jwt, converter)
+  return AuthAwareAuthenticationToken(jwt, principal, authorities)
+}
+
+private fun findPrincipal(claims: Map<String, Any?>): String {
+  return if (claims.containsKey("user_name")) {
```

Comment:
I've rewritten this since this version - see https://github.com/ministryofjustice/hmpps-prisoner-download-api/blob/da873917e0ee0b24b0a3fa8fe86552ebd8fd5f57/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/AuthAwareTokenConverter.kt